pennylane/plugins/tests/__init__.py

```diff
@@ -47,4 +32,205 @@
+            values and probability. Only takes affect if ``analytic=False``. If not
+            provided, the device default is used.
+        skip_ops (bool): whether to skip tests that use operations not supported
+            by the device
+        pytest_args (list[str]): additional PyTest arguments and flags
+        **kwargs: Additional device keyword args
+
+    **Example**
+
+    >>> from pennylane.plugins.tests import test_device
+    >>> test_device("default.qubit")
+    ================================ test session starts =======================================
+    platform linux -- Python 3.7.7, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
+    rootdir: /home/josh/xanadu/pennylane/pennylane/plugins/tests, inifile: pytest.ini
+    plugins: flaky-3.6.1, cov-2.8.1, mock-3.1.0
+    collected 86 items
+    xanadu/pennylane/pennylane/plugins/tests/test_gates.py ..............................
+    ...............................                                                       [ 70%]
+    xanadu/pennylane/pennylane/plugins/tests/test_measurements.py .......sss...sss..sss   [ 95%]
+    xanadu/pennylane/pennylane/plugins/tests/test_properties.py ....                      [100%]
+    ================================= 77 passed, 9 skipped in 0.78s ============================
+
+    """
+    try:
+        import pytest  # pylint: disable=unused-import
+        import pytest_mock  # pylint: disable=unused-import
+        import flaky  # pylint: disable=unused-import
+    except ImportError:
+        raise ImportError(
+            "The device tests requires the following Python packages:"
+            "\npytest pytest-cov pytest_mock flaky"
+            "\nThese can be installed using pip."
+        )
+
+    pytest_args = pytest_args or []
+    test_dir = get_device_tests()
+
+    cmds = ["pytest"]
+    cmds.append(test_dir)
+    cmds.append(f"--device={device}")
+
+    if shots is not None:
+        cmds.append(f"--shots={shots}")
+
+    if analytic is not None:
+        cmds.append(f"--analytic={analytic}")
+
+    if skip_ops:
+        cmds.append("--skip-ops")
+
+    if not flaky_report:
+        cmds.append("--no-flaky-report")
+
+    if kwargs:
+        device_kwargs = " ".join([f"{k}={v}" for k, v in kwargs.items()])
+        cmds += ["--device-kwargs", device_kwargs]
+
+    try:
+        subprocess.run(cmds + pytest_args, check=not interactive)
+    except subprocess.CalledProcessError as e:
+        # pytest return codes:
+        #   Exit code 0:    All tests were collected and passed successfully
+        #   Exit code 1:    Tests were collected and run but some of the tests failed
+        #   Exit code 2:    Test execution was interrupted by the user
+        #   Exit code 3:    Internal error happened while executing tests
+        #   Exit code 4:    pytest command line usage error
+        #   Exit code 5:    No tests were collected
+        if e.returncode in range(1, 6):
+            sys.exit(1)
+        raise e
```

Comment:
Just for my own understanding: if code 1-5 were received, is then the corresponding error message propagated to the parent process (stored in the `returncode` of the object)?

Reply:
No, I found that if error code 1-5 occur, the `CalledProcessError` would re-raise the already raised PyTest error. So the user would see it twice.

This is just to avoid a known error message from being raised twice

Reply:
Have added a comment making it clearer :)